lib/sparse/src/index/posting_list.rs

```diff
@@ -29,6 +29,18 @@ impl PostingList {
         }
     }
 
+    pub fn delete(&mut self, record_id: PointOffsetType) {
+        let index = self
+            .elements
+            .binary_search_by_key(&record_id, |e| e.record_id);
+        if let Ok(found_index) = index {
+            self.elements.remove(found_index);
+            if found_index < self.elements.len() {
+                self.propagate_max_next_weight_to_the_left(found_index);
+            }
+        }
+    }
+
```

Comment:
Randomized test show that resulting `max_next_weight`s are not the same as if we recreate the same index from scratch without the deleted element.

But since the resulting `max_next_weight`s are greater or equal than the proper values, it'll not break the pruning logic, but may make it less efficient.

Not sure if it worth fixing giving that we probably get rid of of the pruning soon. But maybe add a commend and a test checking that `max_next_weight[i] >= max(weight[i..])`.

<details>
<summary>quick and dirty test code</summary>

```rust
#[test]
fn test_random_delete() {
    use rand::Rng;
    use rand::seq::SliceRandom;
    let mut rng = rand::thread_rng();
    for _ in 0..1000 {
        let mut ids = Vec::new();
        let mut cur_id = 0;
        for _ in 0..32 {
            cur_id += rng.gen_range(1..10);
            ids.push(cur_id);
        }
        ids.shuffle(&mut rng);
        let random_id = ids[rng.gen_range(0..ids.len())];

        let mut builder1 = PostingBuilder::new();
        let mut builder2 = PostingBuilder::new();
        for id in ids {
            let val = rng.gen_range(0..100) as f32 / 10.0;
            builder1.add(id, val);
            if id != random_id {
                builder2.add(id, val);
            }
        }

        let mut posting_list1 = builder1.build();
        posting_list1.delete(random_id);
        let posting_list2 = builder2.build();

        // Ok
        assert_eq!(
            posting_list1.elements.iter().map(|e| e.record_id).collect_vec(),
            posting_list2.elements.iter().map(|e| e.record_id).collect_vec(),
        );
        assert_eq!(
            posting_list1.elements.iter().map(|e| e.weight).collect_vec(),
            posting_list2.elements.iter().map(|e| e.weight).collect_vec(),
        );

        // Fail
        assert_eq!(
            posting_list1.elements.iter().map(|e| e.max_next_weight).collect_vec(),
            posting_list2.elements.iter().map(|e| e.max_next_weight).collect_vec(),
        );

        // Ok (at least they won't break pruning logic)
        assert!(
            std::iter::zip(
                &posting_list1.elements,
                &posting_list2.elements,
            ).all(|(e1, e2)| e1.max_next_weight >= e2.max_next_weight),
        );
    }
}
```

</details>

Reply:
Fixed the correctness, but the code doesn't look nice.

Let's remove this heuristic in the next PR